File: soungmunkim/Python/Algorithms/SingleShortestPath_Dijkstra.py
"""
[Dijkstra algorithm 구현하기]

negative weight cycle 없다는 가정
"""

import heapq  # 우선순위 큐를 위한 heapq 모듈을 사용합니다.

class Vertex:
    def __init__(self, name):
        self.name = name
        self.d = float('inf')  # v와 s 사이의 거리를 나타냅니다.
        self.pi = None  # v의 부모 노드를 나타냅니다.

    def __lt__(self, other):
        return self.d < other.d
    
    def __str__(self):
        return self.name


class Graph:
    def __init__(self, V, E, Adj=None):
        self.V = V  # 정점의 집합
        self.E = E  # 간선의 집합
        self.Adj = Adj  # 인접 리스트

def INITIALIZE_SINGLE_SOURCE(G, s):
    for v in G.V:
        v.d = float('inf')  # v와 s 사이의 거리를 무한대로 설정합니다.
        v.pi = None  # v의 부모 노드를 None으로 설정합니다.
    s.d = 0  # 시작 정점 s와 s 사이의 거리는 0입니다.

def RELAX(u, v, weights):
    if (u, v) in weights:
        if v.d > u.d + weights[(u, v)]:
            v.d = u.d + weights[(u, v)]
            v.pi = u

def DIJKSTRA(G, weights, s):
    """
    그래프 G에서 주어진 시작 정점 s로부터의 최단 경로를 계산합니다.
    
    :param G: 그래프
    :param weights: 간선의 가중치를 나타내는 딕셔너리
    :param s: 시작 정점
    """
    INITIALIZE_SINGLE_SOURCE(G, s)
    S = set()  # 경로를 저장하는 컨테이너
    Q = [(v.d, v) for v in G.V]  # (거리, 정점) 형태의 튜플로 우선순위 큐를 초기화합니다.
    heapq.heapify(Q)  # Min-priority queue로 만듭니다.

    while Q:
        _, u = heapq.heappop(Q)  # s와의 거리가 최소인 정점 u를 추출합니다.
        S.add(u)  # u를 S에 추가합니다.
        for v in G.Adj[u]:
            old_distance = v.d
            RELAX(u, v, weights)
            # RELAX 호출로 인해 v.d가 감소한 경우, 우선순위 큐에서 v의 키를 갱신합니다.
            if v.d < old_distance:
                # Q를 리스트로 처리하므로, v의 위치를 찾아서 갱신해야 합니다.
                # 실제 응용에서는 더 효율적인 자료구조를 사용해야 합니다.
                for i, (d, vertex) in enumerate(Q):
                    if vertex == v:
                        Q[i] = (v.d, v)
                        break
                heapq.heapify(Q)  # 힙 속성을 유지하기 위해 다시 heapify 합니다.

if __name__ == "__main__":
    # 그래프 초기화
    v1 = Vertex("v1")
    v2 = Vertex("v2")
    v3 = Vertex("v3")
    v4 = Vertex("v4")

    G = Graph([v1, v2, v3, v4], [(v1, v2), (v1, v3), (v2, v4), (v3, v4)])
    G.Adj = {
        v1: [v2, v3],
        v2: [v4],
        v3: [v4],
        v4: []
    }
    weights = {(v1, v2): 5, (v1, v3): 3, (v2, v4): 2, (v3, v4): 1}

    # DIJKSTRA 실행
    DIJKSTRA(G, weights, v1)

    # 결과 출력
    for vertex in G.V:
        print(f"Distance from {v1} to {vertex}: {vertex.d}")

File: soungmunkim/C/Algorithms/SingleShortestPath_Dijkstra.c
/*
[Dijkstra algorithm 구현하기]

negative weight cycle 없다는 가정
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

// Vertex라는 이름의 구조체 타입 정의
// 각 정점은 아이디, 시작 정점으로부터의 거리, 부모 정점 정보를 가지고 있다.
typedef struct Vertex {
    int id;               // 각 정점의 고유한 번호 (아이디)
    double d;             // 시작 정점에서 현재 정점까지의 최단 거리
    struct Vertex* pi;   // 최단 경로 트리에서 현재 정점의 부모 정점
} Vertex;

// Graph라는 이름의 구조체 타입 정의
typedef struct Graph {
    Vertex** V;               // 그래프의 모든 정점을 저장하는 포인터 배열
    int numVertices;         // 그래프의 정점 개수
    Vertex*** Adj;           // 그래프의 각 정점에 연결된 인접 정점 리스트
    int* numAdjVertices;     // 각 정점의 인접 정점 개수를 저장하는 배열
} Graph;

// 시작 정점을 받아서 모든 정점의 거리를 무한대로 초기화하고 시작 정점만 0으로 설정
void INITIALIZE_SINGLE_SOURCE(Graph* G, Vertex* s) {
    for (int i = 0; i < G->numVertices; i++) {
        G->V[i]->d = INT_MAX;   // 최대 값으로 거리 초기화
        G->V[i]->pi = NULL;    // 부모 정점 초기화
    }
    s->d = 0;   // 시작 정점의 거리는 0
}

// 두 정점과 가중치 정보를 받아서 필요한 경우 경로를 갱신 (Relaxation)
void RELAX(Vertex* u, Vertex* v, int weights[][2]) {
    int weight = weights[u->id][v->id];
    if (v->d > u->d + weight) {
        v->d = u->d + weight;  // 거리 갱신
        v->pi = u;             // 부모 정점 갱신
    }
}

// 정점들을 거리를 기준으로 정렬하는 함수
int compare(const void* a, const void* b) {
    Vertex* vertexA = *(Vertex**)a;
    Vertex* vertexB = *(Vertex**)b;
    if (vertexA->d < vertexB->d) return -1;
    if (vertexA->d > vertexB->d) return 1;
    return 0;
}

// 경로를 출력하는 함수
void printPath(Vertex* s, Vertex* v) {
    if (v == s) {
        printf("v%d ", s->id + 1);
    } else if (v->pi == NULL) {
        printf("no path from v%d to v%d exists", s->id + 1, v->id + 1);
    } else {
        printPath(s, v->pi);
        printf("-> v%d ", v->id + 1);
    }
}

// Dijkstra 알고리즘 구현 함수
void DIJKSTRA(Graph* G, int weights[][2], Vertex* s) {
    // 모든 정점의 거리와 부모를 초기화
    INITIALIZE_SINGLE_SOURCE(G, s);
    Vertex** S = NULL;   // 방문한 정점을 저장할 배열
    int numS = 0;        // 방문한 정점의 개수
    Vertex** Q = (Vertex**)malloc(G->numVertices * sizeof(Vertex*));  // 아직 방문하지 않은 정점들의 배열
    for (int i = 0; i < G->numVertices; i++) {
        Q[i] = G->V[i];  // 초기에는 모든 정점이 방문되지 않았으므로 Q에 추가
    }
    int numQ = G->numVertices;

    while (numQ > 0) {  // 모든 정점을 방문할 때까지 반복
        qsort(Q, numQ, sizeof(Vertex*), compare);  // Q 배열을 거리를 기준으로 정렬
        Vertex* u = Q[0];  // 현재 가장 거리가 짧은 정점 선택
        S = (Vertex**)realloc(S, (numS + 1) * sizeof(Vertex*));  // S 배열에 추가
        S[numS++] = u;
        for (int i = 0; i < G->numAdjVertices[u->id]; i++) {  // 선택한 정점과 연결된 모든 정점에 대해
            Vertex* v = G->Adj[u->id][i];
            RELAX(u, v, weights);  // Relaxation 실행
        }
        for (int i = 0; i < numQ - 1; i++) {
            Q[i] = Q[i + 1];  // 현재 정점을 Q에서 제거
        }
        numQ--;  // Q의 크기 감소
    }
    free(Q);
    free(S);
}


int main() {
    // 그래프 초기화 부분
    Vertex v1 = {0, INT_MAX, NULL};
    Vertex v2 = {1, INT_MAX, NULL};
    Vertex v3 = {2, INT_MAX, NULL};
    Vertex v4 = {3, INT_MAX, NULL};

    Vertex* adj_v1[] = {&v2, &v3};
    Vertex* adj_v2[] = {&v4};
    Vertex* adj_v3[] = {&v4};

    Graph G = {
        .V = (Vertex*[]){&v1, &v2, &v3, &v4},
        .numVertices = 4,
        .Adj = (Vertex**[]){adj_v1, adj_v2, adj_v3, NULL},
        .numAdjVertices = (int[]){2, 1, 1, 0}
    };

    int weights[4][2] = {  // 각 간선의 가중치 정보
        {0, 5},
        {0, 3},
        {1, 2},
        {2, 1}
    };

    // Dijkstra 알고리즘을 실행하여 최단 경로 계산
    DIJKSTRA(&G, weights, &v1);

    // 결과 출력
    for (int i = 0; i < G.numVertices; i++) {
        printf("Distance from v1 to v%d: %f\n", G.V[i]->id + 1, G.V[i]->d);
        printf("Path: ");
        printPath(&v1, G.V[i]);
        printf("\n");
    }
}
